### PR TITLE
Patch various classes to allow modded bows and crossbows to work correctly

### PR DESCRIPTION
**Pull request by Thelnfamous1**
Builds on the code from [#6606](https://github.com/MinecraftForge/MinecraftForge/pull/6606)

A new version of getHandWith in ProjectileHelper is added that takes in an item Predicate instead of an Item. This new version replaces all vanilla usages of the original, making it deprecated. Several ranged mobs use the original getHandWith to coordinate their ranged attacks, and the hardcoded Item check involved prevents them from using modded versions of their associated ranged weapon. With the new version of getHandWith, they can use modded ranged weapons without issues.

In addition to that, there are several checks for "== Items.BOW" and "== Items.CROSSBOW" that were replaced with checks for "instanceof BowItem" and "instanceof CrossbowItem" respectively to allow for complete functionality of modded bows and crossbows. 

Included is a test mod that adds a modded bow and crossbow. Here are some commands for testing their usages by the Skeleton, Pillager, Illusioner and Piglin.

/summon minecraft:skeleton ~ ~ ~ {HandItems:[{Count:1,id:"get_hand_with_test:modded_bow"},{}]}

/summon minecraft:pillager ~ ~ ~ {HandItems:[{Count:1,id:"get_hand_with_test:modded_crossbow"},{}]}

/summon minecraft:illusioner ~ ~ ~ {HandItems:[{Count:1,id:"get_hand_with_test:modded_bow"},{}]}

/summon minecraft:piglin ~ ~ ~ {HandItems:[{Count:1,id:"get_hand_with_test:modded_crossbow"},{}]}

### CODE REVIEW

File: patches/minecraft/net/minecraft/entity/projectile/ProjectileHelper.java.patch [hunk 1]
```diff
@@ -9,3 +9,19 @@
                    if (d0 == 0.0D) {
                       entity = entity1;
                       vector3d = vector3d1;
+@@ -123,10 +123,15 @@
+       }
+    }
+ 
++   @Deprecated
```
Comment (pupnewfster) [inline patches/minecraft/net/minecraft/entity/projectile/ProjectileHelper.java.patch]: You probably should add a comment so then this line becomes like:
```java
@Deprecated //Forge: Use the predicate version of getHandWidth instead
```

File: patches/minecraft/net/minecraft/entity/player/ServerPlayerEntity.java.patch [hunk 2]
```diff
@@ -130,23 +130,15 @@
     }
  
     protected void func_70670_a(EffectInstance p_70670_1_) {
-@@ -1139,6 +1160,7 @@
-    }
- 
-    public void func_71033_a(GameType p_71033_1_) {
-+      if (!net.minecraftforge.common.ForgeHooks.onChangeGameMode(this, this.field_71134_c.func_73081_b(), p_71033_1_)) return;
```
Comment (pupnewfster) [inline patches/minecraft/net/minecraft/entity/player/ServerPlayerEntity.java.patch]: Not sure why this line is being removed?

File: patches/minecraft/net/minecraft/client/network/play/NetworkPlayerInfo.java.patch [hunk 3]
```diff
@@ -1,10 +0,0 @@
---- a/net/minecraft/client/network/play/NetworkPlayerInfo.java
-+++ b/net/minecraft/client/network/play/NetworkPlayerInfo.java
-@@ -50,6 +50,7 @@
-    }
- 
-    protected void func_178839_a(GameType p_178839_1_) {
-+      net.minecraftforge.client.ForgeHooksClient.onClientChangeGameMode(this, this.field_178866_b, p_178839_1_);
```
Comment (pupnewfster) [inline patches/minecraft/net/minecraft/client/network/play/NetworkPlayerInfo.java.patch]: Not sure why this line is being removed?

File: patches/minecraft/net/minecraft/entity/player/ServerPlayerEntity.java.patch (hunk 2 above)
Comment (Thelnfamous1) [reply patches/minecraft/net/minecraft/entity/player/ServerPlayerEntity.java.patch]: I'm not sure why that happened either, I didn't touch that

File: patches/minecraft/net/minecraft/client/network/play/NetworkPlayerInfo.java.patch (hunk 3 above)
Comment (Thelnfamous1) [reply patches/minecraft/net/minecraft/client/network/play/NetworkPlayerInfo.java.patch]: Again, not sure why that was also touched

File: patches/minecraft/net/minecraft/entity/projectile/ProjectileHelper.java.patch (hunk 1 above)
Comment (Thelnfamous1) [reply patches/minecraft/net/minecraft/entity/projectile/ProjectileHelper.java.patch]: Will do

**Comment by Thelnfamous1** [comment]
I'm not sure how to go about reverting the changes made to ServerPlayerEntity and NetworkPlayerInfo classes, as I didn't mean those to be a part of this PR at all

**Comment by pupnewfster** [comment]
I am not fully sure but I believe at least in theory if you add those lines back and then rerun genPatches it probably will end up "reverting" those changes.

**Comment by Thelnfamous1** [comment]
Removed the unintentional patch changes, every file listed under "Changes" should now be the ones that are a part of this patch

**Comment by Thelnfamous1** [comment]
If there's anything else I need to do on my end to get this merged, I'd appreciate any input.

**Comment by LexManos** [comment]
So, as discussed on discord, this isn't going to go in in its current state.
This is just too widespread of changes for little effect.
The maintenance cost is a nightmare.
However, as a proof of concept i've worked on this: https://github.com/LexManos/MinecraftForge/commit/9ffa44f254d9d9de017feaee37780e0707b9a07a
A system for semi-data driven transformers to be used in Forge.
The basic idea is that all these simple patches that are just the same thing, 50 times, would be done via a coremod on our end.
Combined with build script changes to verify and detect the targets so that when Minecraft updates it's just a matter of regenning the data files.

If you wanna try your hand the linked branch is my WIP that's very low priority.

**Comment by TheCurle** [comment]
How is this looking with the field to instanceof transformers we package?

**Pull request closed without merging**